File: src/types/Cantrip.ts
import {
  Array,
  Boolean,
  Entity,
  Enum,
  EnumCase,
  IncludeIdentifier,
  Integer,
  Object,
  Optional,
  Required,
  String,
  TypeAlias,
} from "tsondb/schema/dsl"
import { CastingTimeDuringLovemaking } from "./_ActivatableSkillCastingTime.js"
import { DurationUnit } from "./_ActivatableSkillDuration.js"
import { FixedRange } from "./_ActivatableSkillRange.js"
import { AffectedTargetCategories } from "./_ActivatableSkillTargetCategory.js"
import {
  CurriculumIdentifier,
  MagicalTraditionIdentifier,
  PropertyIdentifier,
} from "./_Identifier.js"
import { ResponsiveText } from "./_ResponsiveText.js"
import { NestedTranslationMap } from "./Locale.js"
import { Errata } from "./source/_Erratum.js"
import { src } from "./source/_PublicationRef.js"

export const Cantrip = Entity(import.meta.url, {
  name: "Cantrip",
  namePlural: "Cantrips",
  type: () =>
    Object({
      parameters: Required({
        comment: "Measurable parameters of a cantrip.",
        type: IncludeIdentifier(CantripPerformanceParameters),
      }),
      target: Required({
        comment: "The target category – the kind of creature or object – the skill affects.",
        type: IncludeIdentifier(AffectedTargetCategories),
      }),
      property: Required({
        comment: "The associated property.",
        type: PropertyIdentifier(),
      }),
      note: Optional({
        comment:
          "A note specifying the dissemination of the cantrip in different traditions. Sometimes a cantrip is exclusively available to one or more specific traditions, but usually one the academies and traditions are listed the cantrip is most commonly teached in.",
        type: IncludeIdentifier(CantripNote),
      }),
      src,
      translations: NestedTranslationMap(
        Required,
        "Cantrip",
        Object({
          name: Required({
            comment: "The cantrip’s name.",
            type: String({ minLength: 1 }),
          }),
          effect: Required({
            comment: "The effect description.",
            type: String({ minLength: 1, isMarkdown: true }),
          }),
          range: Required({
            isDeprecated: true,
            type: String({ minLength: 1 }),
          }),
          duration: Required({
            isDeprecated: true,
            type: String({ minLength: 1 }),
          }),
          target: Required({
            isDeprecated: true,
            type: String({ minLength: 1 }),
          }),
          errata: Optional({
            type: IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

const CantripNote = Enum(import.meta.url, {
  name: "CantripNote",
  comment:
    "A note specifying the dissemination of the cantrip in different traditions. Sometimes a cantrip is exclusively available to one or more specific traditions, but usually one the academies and traditions are listed the cantrip is most commonly teached in.",
  values: () => ({
    Exclusive: EnumCase({ type: IncludeIdentifier(ExclusiveCantripNote) }),
    Common: EnumCase({ type: IncludeIdentifier(CommonCantripNotes) }),
  }),
})

const ExclusiveCantripNote = TypeAlias(import.meta.url, {
  name: "ExclusiveCantripNote",
  type: () =>
    Object({
      traditions: Required({
        comment: "The traditions the cantrip is exclusively available to.",
        type: Array(MagicalTraditionIdentifier(), { minItems: 1, uniqueItems: true }),
      }),
    }),
})

const CommonCantripNotes = TypeAlias(import.meta.url, {
  name: "CommonCantripNotes",
  type: () =>
    Object({
      list: Required({
        comment: "The academies and traditions the cantrip is commonly teached in.",
        type: Array(IncludeIdentifier(CommonCantripNote), { minItems: 1, uniqueItems: true }),
      }),
    }),
})

const CommonCantripNote = Enum(import.meta.url, {
  name: "CommonCantripNote",
  values: () => ({
    Academy: EnumCase({ type: CurriculumIdentifier() }),
    Tradition: EnumCase({ type: IncludeIdentifier(CommonCantripTraditionNote) }),
  }),
})

const CommonCantripTraditionNote = TypeAlias(import.meta.url, {
  name: "CommonCantripTraditionNote",
  type: () =>
    Object({
      id: Required({
        comment: "The magical tradition’s identifier.",
        type: MagicalTraditionIdentifier(),
      }),
      translations: NestedTranslationMap(
        Optional,
        "CommonCantripTraditionNote",
        Object({
          note: Required({
            comment: "A note, appended to the generated string in parenthesis.",
            type: String({ minLength: 1 }),
          }),
        }),
      ),
    }),
})

const CantripPerformanceParameters = TypeAlias(import.meta.url, {
  name: "CantripPerformanceParameters",
  comment: "Measurable parameters of a cantrip.",
  type: () =>
    Object({
      range: Required({ type: IncludeIdentifier(CantripRange) }),
      duration: Required({ type: IncludeIdentifier(CantripDuration) }),
    }),
})

const CantripRange = Enum(import.meta.url, {
  name: "CantripRange",
  values: () => ({
    Self: EnumCase({ type: null }),
    Touch: EnumCase({ type: null }),
    Fixed: EnumCase({ type: IncludeIdentifier(FixedRange) }),
  }),
})

const CantripDuration = Enum(import.meta.url, {
  name: "CantripDuration",
  values: () => ({
    Immediate: EnumCase({ type: null }),
    Fixed: EnumCase({ type: IncludeIdentifier(FixedCantripDuration) }),
    DuringLovemaking: EnumCase({ type: IncludeIdentifier(CastingTimeDuringLovemaking) }),
    Indefinite: EnumCase({ type: IncludeIdentifier(IndefiniteCantripDuration) }),
  }),
})

const FixedCantripDuration = TypeAlias(import.meta.url, {
  name: "FixedCantripDuration",
  type: () =>
    Object({
      is_maximum: Optional({
        comment: "If the duration is the maximum duration, so it may end earlier.",
        type: Boolean(),
      }),
      value: Required({
        comment: "The (unitless) duration.",
        type: Integer({ minimum: 1 }),
      }),
      unit: Required({
        comment: "The duration unit.",
        type: IncludeIdentifier(DurationUnit),
      }),
    }),
})

const IndefiniteCantripDuration = TypeAlias(import.meta.url, {
  name: "IndefiniteCantripDuration",
  type: () =>
    Object({
      translations: NestedTranslationMap(
        Required,
        "IndefiniteCantripDuration",
        Object({
          description: Required({
            comment: "A description of the duration.",
            type: IncludeIdentifier(ResponsiveText),
          }),
        }),
      ),
    }),
})

File: Sources/OptolithDatabaseSchema/Types/Cantrip.swift
import FileDB

@Model
public struct Cantrip {

  /// Measurable parameters of a cantrip.
  let parameters: CantripPerformanceParameters

  /// The target category – the kind of creature or object – the skill affects.
  let target: AffectedTargetCategories

  /// The associated property.
  @Relationship(Property.self)
  let property: Property.ID

  /// A note specifying the dissemination of the cantrip in different traditions. Sometimes a cantrip is exclusively available to one or more specific traditions, but usually one the academies and traditions are listed the cantrip is most commonly teached in.
  let note: CantripNote?

    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation { // CantripTranslation

        /// The cantrip’s name.
        @MinLength(1)
        let name: String

        /// The effect description.
        @MinLength(1)
        @Markdown
        let effect: String

        @available(*, deprecated, message: "Use language-independent performance parameters instead")
        @MinLength(1)
        let range: String

        @available(*, deprecated, message: "Use language-independent performance parameters instead")
        @MinLength(1)
        let duration: String

        @available(*, deprecated, message: "Use language-independent performance parameters instead")
        @MinLength(1)
        let target: String

        /// A list of errata for the entry in the specific language.
        @MinItems(1)
        let errata: [Erratum]?
    }
}

/// A note specifying the dissemination of the cantrip in different traditions. Sometimes a cantrip is exclusively available to one or more specific traditions, but usually one the academies and traditions are listed the cantrip is most commonly teached in.
@ModelEnum
public enum CantripNote {
    case Exclusive(ExclusiveCantripNote)
    case Common(CommonCantripNotes)
}

@Embedded
public struct ExclusiveCantripNote {

  /// The traditions the cantrip is exclusively available to.
  @MinItems(1)
  @UniqueItems
  @Relationship(MagicalTradition.self)
  let traditions: [MagicalTradition.ID]
  }

@Embedded
public struct CommonCantripNotes {

  /// The academies and traditions the cantrip is commonly teached in.
  @MinItems(1)
  @UniqueItems
  let list: [CommonCantripNote]
  }

@ModelEnum
public enum CommonCantripNote {
    case Academy(CurriculumIdentifierObject)
    case Tradition(CommonCantripTraditionNote)
}

@Embedded
public struct CommonCantripTraditionNote {

  /// The magical tradition’s identifier.
  @Relationship(MagicalTradition.self)
  let id: MagicalTradition.ID

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation { // CommonCantripTraditionNoteTranslation

        /// A note, appended to the generated string in parenthesis.
        @MinLength(1)
        let note: String
      }
  }

/// Measurable parameters of a cantrip.
@Embedded
public struct CantripPerformanceParameters {
      let range: CantripRange
      let duration: CantripDuration
  }

@ModelEnum
public enum CantripRange {
    case `Self`
    case Touch
    case Fixed(FixedRange)
}

@ModelEnum
public enum CantripDuration {
    case Immediate
    case Fixed(FixedCantripDuration)
    case DuringLovemaking(CastingTimeDuringLovemaking)
    case Indefinite(IndefiniteCantripDuration)
}

@Embedded
public struct FixedCantripDuration {
  /// If the duration is the maximum duration, so it may end earlier.
  let is_maximum: Bool?

  /// The (unitless) duration.
  @Minimum(1)
  let value: Int

  /// The duration unit.
  let unit: DurationUnit
  }

@Embedded
public struct IndefiniteCantripDuration {

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation { // IndefiniteCantripDurationTranslation

        /// A description of the duration.
        let description: ResponsiveText
      }
  }
